--- uboot/arch/arm/mach-aspeed/board.c	2020-12-04 17:55:03.650268424 +0530
+++ uboot_new/arch/arm/mach-aspeed/board.c	2020-12-04 19:46:29.519266092 +0530
@@ -265,10 +265,24 @@
 int board_init(void)
 {
 	struct udevice *dev;
-
+	int i;
+	int ret;
 	gd->bd->bi_boot_params = CONFIG_SYS_SDRAM_BASE + 0x100;
 
-	uclass_first_device_check(UCLASS_MISC, &dev);
+	/*
+	 * Loop over all MISC uclass drivers to call the comphy code
+	 * and init all CP110 devices enabled in the DT
+	 */
+	i = 0;
+	while (1) {
+		/* Call the comphy code via the MISC uclass driver */
+		ret = uclass_get_device(UCLASS_MISC, i++, &dev);
+
+		/* We're done, once no further CP110 device is found */
+		if (ret)
+			break;
+	}
+
 
 #if defined(CONFIG_EARLYBOOT_ESPI_HANDSHAKE) 
 	u32 strap;    
